data_safe_haven/pulumi/components/sre_dns_server.py

```diff
@@ -0,0 +1,319 @@
+        resources_path = (
+            pathlib.Path(__file__).parent.parent.parent / "resources" / "dns_server"
+        )
+        adguard_entrypoint_sh_reader = FileReader(resources_path / "entrypoint.sh")
+        adguard_adguardhome_yaml_reader = FileReader(
+            resources_path / "AdGuardHome.mustache.yaml"
+        )
+
+        # Expand AdGuardHome YAML configuration
+        adguard_adguardhome_yaml_contents = Output.all(
+            admin_username=props.admin_username,
+            admin_password_encrypted=props.admin_password.apply(
+                lambda p: bcrypt_encode(p)
+            ),
+            upstream_dns="168.63.129.16",
+            filter_deny=["*.*"],
+            filter_allow=Output.from_input(props.shm_fqdn).apply(
+                lambda fqdn: [f"*.{fqdn}", *allowed_dns_lookups()]
+            ),
+        ).apply(
+            lambda mustache_values: adguard_adguardhome_yaml_reader.file_contents(
+                mustache_values
+            )
+        )
+
+        # Define network security group
+        nsg = network.NetworkSecurityGroup(
+            f"{self._name}_nsg_dns",
+            network_security_group_name=f"{stack_name}-nsg-dns",
+            resource_group_name=resource_group.name,
+            security_rules=[
+                # Inbound
+                network.SecurityRuleArgs(
+                    access=network.SecurityRuleAccess.ALLOW,
+                    description="Allow inbound connections from attached.",
+                    destination_address_prefix=props.ip_range_prefix,
+                    destination_port_ranges=["53"],
+                    direction=network.SecurityRuleDirection.INBOUND,
+                    name="AllowSREInbound",
+                    priority=NetworkingPriorities.INTERNAL_SRE_ANY,
+                    protocol=network.SecurityRuleProtocol.ASTERISK,
+                    source_address_prefix=props.sre_vnet_prefix,
+                    source_port_range="*",
+                ),
+                network.SecurityRuleArgs(
+                    access=network.SecurityRuleAccess.DENY,
+                    description="Deny all other inbound traffic.",
+                    destination_address_prefix="*",
+                    destination_port_range="*",
+                    direction=network.SecurityRuleDirection.INBOUND,
+                    name="DenyAllOtherInbound",
+                    priority=NetworkingPriorities.ALL_OTHER,
+                    protocol=network.SecurityRuleProtocol.ASTERISK,
+                    source_address_prefix="*",
+                    source_port_range="*",
+                ),
+                # Outbound
+                network.SecurityRuleArgs(
+                    access=network.SecurityRuleAccess.ALLOW,
+                    description="Allow outbound DNS and rules list traffic over the internet.",
+                    destination_address_prefix="Internet",
+                    destination_port_ranges=["53", "80", "443"],
```

Comment:
Does this need HTTP and HTTPS as well?

Reply:
Originally it did, as AdGuard supports use of remote block lists which are retrieved over HTTP(S). I decided that maintaining our own block list that had to always be available in the same location (e.g. on GitHub) wasn't very future proof. I think we can lose 80 and 443 here in that case.

Reply:
Done in 1860e930